src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java

```diff
@@ -4350,6 +4224,135 @@ && serverFeaturesModified()) {
+     * @param numApplicationUpdatedMessages number of application updated
+     *                                      messages, used when running tests
+     * @param configuredServerXml           true if fileChanged is serverXmlFile
+     *                                      specified by the configuration parameter
+     * @throws IOException
+     * @throws PluginExecutionException
+     */
+    private void processConfigFileChange(File fileChanged, ChangeType changeType, ThreadPoolExecutor executor,
+            int numApplicationUpdatedMessages, boolean configuredServerXml)
+            throws IOException, PluginExecutionException {
+        boolean isGeneratedFeaturesFile = configuredServerXml ? false : fileChanged.equals(generatedFeaturesFile);
+        String targetFileName = configuredServerXml ? "server.xml" : null; // if null file will retain the same name when copied
+        File fileChangedParentDir = configuredServerXml ? serverXmlFileParent : configDirectory;
+
+        if (fileChanged.exists() && (changeType == ChangeType.MODIFY || changeType == ChangeType.CREATE)) {
+            debug("Config file modified: " + fileChanged);
+            boolean generateFeaturesSuccess = true; // default to true in case feature generation is disabled
+            boolean serverFeaturesModified = serverFeaturesModified();
+
+            // generate features whenever features have changed and an XML file is modified,
+            // excluding the generated features file
+            if (generateFeatures && (fileChanged.getName().endsWith(".xml")
+                    && !isGeneratedFeaturesFile)
+                    && serverFeaturesModified) {
+                generateFeaturesSuccess = optimizeGenerateFeatures();
+            }
+            if (serverFeaturesModified) {
+                // suppress install feature warning - property must be set before calling
+                // installFeaturesToTempDir
+                System.setProperty(SKIP_BETA_INSTALL_WARNING, Boolean.TRUE.toString());
+                installFeaturesToTempDir(fileChanged, fileChangedParentDir, targetFileName, generateFeaturesSuccess);
+            }
+            copyFile(fileChanged, fileChangedParentDir, serverDirectory, targetFileName);
+
+            // if the generated features file was modified as a result of another config
+            // file modification, copy it over to target so the server picks up the changes
+            // together
+            if (generateFeaturesSuccess && generatedFeaturesModified && !isGeneratedFeaturesFile) {
+                // this logic is not entered if the fileChanged is the generated features file
+                // copy generated features file to server dir
+                copyFile(generatedFeaturesFile, configDirectory, serverDirectory, null);
+                generatedFeaturesModified = false;
+            }
+            if (serverFeaturesModified) {
+                updateExistingFeatures();
```

Comment:
@TrevCraw pointed out that we are updating the existing feature list twice, once here and once in the `installFeatures` logic: 
https://github.com/OpenLiberty/ci.maven/blob/37b62be574e76f25e4e04710bf8d7c6537db5545/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/DevMojo.java#L1010
https://github.com/OpenLiberty/ci.maven/blob/37b62be574e76f25e4e04710bf8d7c6537db5545/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/DevMojo.java#L1003

We should avoid duplicating updating the feature list.

Reply:
Fixed with [Remove redundant logic to update existing features list](https://github.com/OpenLiberty/ci.maven/pull/1541/commits/fb2f9eb8a66ed8fa38fead2f8bbe4477abb66729) and [Remove redundant logic to update existing features list](https://github.com/OpenLiberty/ci.gradle/pull/756/commits/8e76b62d8d9f7d118fa318de9ebcda10daf10963)